core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java

```diff
@@ -2490,4 +2493,80 @@ public void testGetSnapshotsRequest() throws Exception {
         waitForCompletion(repositoryName, inProgressSnapshot, TimeValue.timeValueSeconds(60));
     }
 
+    /**
+     * This test ensures that if a node that holds a primary that is being snapshotted leaves the cluster,
+     * when it returns, the node aborts the snapshotting on the now removed shard.
+     * Before, the node would continue snapshotting, which means it would continue to hold a reference to
+     * the Store for the shard.  Holding this reference would, for example, prevent a replica from being
+     * allocated to the same node, until the snapshot completes.
+     *
+     * See https://github.com/elastic/elasticsearch/issues/20876
+     */
+    public void testSnapshotCanceledOnRemovedShard() throws Exception {
+        final int numPrimaries = 1;
+        final int numReplicas = 1;
+        final int numDocs = 100;
+        final String repo = "test-repo";
+        final String index = "test-idx";
+        final String snapshot = "test-snap";
+
+        assertAcked(prepareCreate(index, 1,
+            Settings.builder().put("number_of_shards", numPrimaries).put("number_of_replicas", numReplicas)));
+
+        logger.info("--> indexing some data");
+        Client client = client();
+        for (int i = 0; i < numDocs; i++) {
+            index(index, "doc", Integer.toString(i), "foo", "bar" + i);
+        }
+        refresh();
+
+        logger.info("--> creating repository");
+        PutRepositoryResponse putRepositoryResponse =
+            client.admin().cluster().preparePutRepository(repo).setType("fs").setSettings(Settings.builder()
+                .put("location", randomRepoPath())
+                // very slow snapshotting, so the node has time to leave the cluster and rejoin while still snapshotting
+                .put("max_snapshot_bytes_per_sec", "100b")
+            ).get();
+        assertTrue(putRepositoryResponse.isAcknowledged());
+
+        logger.info("--> snapshot");
+        client.admin().cluster().prepareCreateSnapshot(repo, snapshot)
+            .setWaitForCompletion(false)
+            .execute();
+
+        logger.info("--> waiting for snapshot to be in progress on all nodes");
+        assertBusy(() -> {
+            for (String node : internalCluster().nodesInclude(index)) {
+                final Client nodeClient = client(node);
```

Comment:
I made a mistake here, this only ensures the snapshot cluster state update has reached master, so I changed it to use `internalCluster().clusterService(node).state()` instead, to ensure each node knows that the snapshot is in progress.